General ballistic aim algorithm improvements

JacekCzupyt opened the pull request:
General ballistic aim algorithm improvements for more accurate targeting, especially using mobile weapons or against high velocity/acceleration targets.

Tested with a modified goalkeeper variant - 0 deviation, 600 rpm, 50km engagement range, default 1109 bullet velocity.
- Can reliably hit static surface targets at 45km.
- Can semi-reliably hit static surface targets while mounted on a maneuvering aircraft at 45km.
- Can reliably hit targets under constant thrust at 45km in low kerbin orbit.

BrettRyland commented:
Your assumptions about the integrator that Unity uses are incorrect. Unity uses PhysX for physics simulations, which is a symplectic integrator. The method you're describing is not symplectic and is thus not sufficiently accurate over the long time-scales required for long-distance ballistics.

Leap-frog is an explicit ¹ second-order symplectic integrator, which is stable for the conditions that we're using it 
for. Since it is symplectic, it allows for much larger time-steps to be used in the simulations (when they're required ²) without significant loss of precision ³.

¹ which is important for performance.
² simulations aren't required if the change in the gravitational field does not change siginificantly over the trajectory, which on Kerbin corresponds to anything less than around 10km, where the deviation is around 20cm.
³ I'm aware of the effects that changing the time-step for the final part of the simulation has on the conservation laws, but the errors accumulated in that final step are sufficiently small as to be irrelevant. I'm also aware of the phase-error inherent in symplectic integrators, but again, it's sufficiently small for these simulations.

Furthermore, and more importantly, BDA bullets are not rigid bodies in Unity ⁴ and are thus not handled by Unity's physics simulation. We perform the numerical integration explicitly using leap-frog, which occurs out-of-sync with Unity's physics. This is also the reason why the "Vessel-Relative Bullet Checks" option exists, which is required to be enabled when the velocities of vessels are sufficient that they can move more than their body length (after accounting for the Krakensbane frame velocity) within a physics frame.

⁴ rockets are however, but their start-up phase also includes significant random variations in the trajectory direction and they're only short-range, so they're not really relevant here.

The current implementation is accurate to within the limits of turrets' ability to stably point in the correct direction and is also capable of hitting a kerbal in the head at 50km.
Some demonstrations of the accuracy of the current implementation:
https://www.youtube.com/watch?v=d6iCj2NBvCI
https://www.youtube.com/watch?v=rW8j5Erp0iY

JacekCzupyt commented:
Regarding the accuracy, while I have seen that the current implementation has basically perfect accuracy for static targets for at least a 50km range, I have not been able to reproduce this accuracy against distant orbital targets, or high acceleration targets (Vessel-Relative Bullet Checks are turned on).

Here some of my tests:
Current implementation - https://youtu.be/6ZO1ZhDqie0
Proposed implementation - https://youtu.be/HQfXLu5shCU

Much to my annoyance, in the tests I setup for the recording, my implementation now achieves a consistent near miss on the 50km target, which I am currently investigating. It's still far closer then I was able to achieve with the current implementation.
Please let me know if anything appears wrong with my testing configuration.

JacekCzupyt commented:
In regards to the numerical methods... it just got a lot more complicated. Turns out we were both half right. 

I've added your patch and successfully replicated the results. I was curious how adding a constant acceleration would effect the outcome, so I modified it by replacing the acceleration from getGeeForceAtPosition to vessel.acceleration_immediate.
<details>
<summary>New test code</summary>

```csharp
// Forward Euler (1st order, not symplectic)
x0 += v0 * Time.fixedDeltaTime; // Update position based on velocity at the current time
v0 += vessel.acceleration_immediate * Time.fixedDeltaTime; // Update the velocity based on the potential at the current time

// Semi-implicit Euler (1st order, symplectic)
v1 += vessel.acceleration_immediate * Time.fixedDeltaTime; // Update velocity based on the potential at the current time
x1 += v1 * Time.fixedDeltaTime; // Update position based on velocity at the future time

// Leap-frog (2nd order, symplectic)
v2 += 0.5f * vessel.acceleration_immediate * Time.fixedDeltaTime; // Update velocity half a step based on the potential at the current time
x2 += v2 * Time.fixedDeltaTime; // Update the position based on the velocity half-way between the current and future times
 v2 += 0.5f * vessel.acceleration_immediate * Time.fixedDeltaTime; // Update velocity another half a step based on the potential at the future time
// Yes I am aware I'm adding the previous steps acceleration value, but for short test lengths, the difference should be negligible 
// Note: combining the second half-update to v2 with the first one of the next step into a single evaluation is where the name of the method comes from.
```
</details>
Yes I am aware that A) I'm cheating by knowing the true acceleration at any given time and B) I'm adding the previous step acceleration in the last leap-frog line, but for testing purposes A isn't an issue, and over short test times, the error from B should be insignificant.

I rerun the test and got near identical results:
```
DEBUG After 2s, Actual x = (0.0, 0.0, 0.0), Forward Euler predicts x = (0.0, 0.0, 0.1) (Δ = 0.1359491), Semi-implicit Euler predicts x = (0.0, 0.0, -0.1) (Δ = 0.1359502), Leap-frog predicts x = (0.0, 0.0, 0.0) (Δ = 0.003325633)
```


However when I tested the results under a constant acceleration, I got the following:


Spacecraft under 5g acceleration
```
DEBUG After 2s, Actual x = (0.0, 0.0, 0.0), Forward Euler predicts x = (1.4, -0.1, -0.8) (Δ = 1.646314), Semi-implicit Euler predicts x = (0.0, 0.0, -0.1) (Δ = 0.1345227), Leap-frog predicts x = (0.7, -0.1, -0.5) (Δ = 0.8728476)
```


Missile under 15g acceleration
```
DEBUG After 2s, Actual x = (-0.1, 0.0, 0.0), Forward Euler predicts x = (5.4, -1.1, -0.9) (Δ = 5.661558), Semi-implicit Euler predicts x = (0.0, 0.0, -0.1) (Δ = 0.1318552), Leap-frog predicts x = (2.7, -0.5, -0.5) (Δ = 2.860085)
```

Both forward euler and leap-frog errors skyrocket, while semi-implicit euler remains basically unchanged, and thus becomes far more accurate the the alternatives.


The reason turns out to the "orbital drift compensation", and option enabled by default in the KSP general settings. It appears that Unity/Physx does in fact Semi-Implicit Euler (supported by [1](https://github.com/NVIDIAGameWorks/PhysX-3.4/issues/57) and [2](https://forums.developer.nvidia.com/t/which-method-does-physx-provide-for-physical-integration/61276), but this option dynamically corrects for the numerical errors it creates exclusively in regards to gravitational forces.

Upon disabling the option and rerunning the tests, Semi-implicit Euler becomes the best estimator.

No manual acceleration:
```
DEBUG After 2s, Actual x = (0.0, 0.0, 0.0), Forward Euler predicts x = (0.0, 0.0, 0.3) (Δ = 0.2719226), Semi-implicit Euler predicts x = (0.0, 0.0, 0.0) (Δ = 0.0001859452), Leap-frog predicts x = (0.0, 0.0, 0.1) (Δ = 0.1359981)
```

Missile at 15g acceleration:
```
DEBUG After 2s, Actual x = (-0.1, 0.0, 0.0), Forward Euler predicts x = (5.4, -1.1, 0.3) (Δ = 5.619738), Semi-implicit Euler predicts x = (-0.1, 0.0, 0.0) (Δ = 0.003372642), Leap-frog predicts x = (2.7, -0.6, 0.1) (Δ = 2.815537)
```


As to how to solve this, option one is to simply disable orbital drift compensation, but that's obviously far from ideal. I will probably try to instead subtract the gravitational acceleration the initial velocity offset.

JacekCzupyt commented on BDArmory/Weapons/ModuleWeapon.cs:
```diff
@@ -2080,9 +2074,9 @@ private void Fire()
+                                    pBullet.timeToLiveUntil = Mathf.Max(maxTargetingRange, maxEffectiveDistance) / bulletVelocity * 1.1f + Time.fixedTime;
 
-                                    timeFired = Time.time - iTime;
+                                    timeFired = Time.fixedTime + Time.fixedDeltaTime - iTime;
```
I was not aware that `Time.time` equivalent to `Time.fixedTime` inside FixedUpdate, I can revert these changes.

However I'm somewhat less confident about the timeFired issue. Components instantiated during a FixedUpdate do not run their own FixedUpdate until the next fixed frame. To verify this, this patch logs the difference between the first fixed update time and time created for bullets:
[Log_first_bullet_fixed_update_time.patch](https://github.com/BrettRyland/BDArmory/files/12363807/Log_first_bullet_fixed_update_time.patch)
There is always a fixedDeltaTime long delay. 

`[BDArmory.PooledBullet] Time between creation and first fixed update: 0.01989746`


My conceptual understanding was that the initial bullet placement represents the bullet being fired in the upcoming physics frame, and compensating for the fact that the bullet will not run its FixedUpdate function before it. Than, every following execution of the bullets FixedUpdate function would represent the bullet simulating its movement during the upcoming physics frame.

If the initial bullet placement is supposed to represent the bullet being fired in the preceding physics frame, that implies that each PooledBullet FixedUpdate function represents the bullet "catching up" to the movement it should make on the preceding physics frame. Otherwise there is "missing frame", where the bullet does not move at all during the physics frame immediately following its creation.

As far as I understand, physics for RigidBody objects happen directly after the FixedUpdate step, I feel like delaying the physics simulation of the bullet until the next FixedUpdate step would be problematic, with potentially multiple Update calls being called in-between, but I could be wrong about this.

What I am confident about however that this is not consistent with the way the PooledBullet FixedUpdate function is structured, specifically the fact that the CheckBulletCollisions check is run **before** the MoveBullet function.

https://github.com/BrettRyland/BDArmory/blob/bfc435d488e66f68b7cb2bf0a792992f590375f5/BDArmory/Ammo/PooledBullet.cs#L422-L424

This would mean bullet collision checks happen between the current physics state of the target, and the bullet physics state from the previous physics frame, which is obviously incorrect.

BrettRyland replied:
Sorry, yes, you're absolutely correct that the `timeFired` is referring to the upcoming frame and so it should have the `Time.fixedDeltaTime` added to it. I'd forgotten about the bullets not being updated on the frame they're enabled.

BrettRyland commented:
Huh, OK, that's very interesting. I was unaware of the Orbital Drift Compensation option in KSP. That would completely mess with any prediction of where objects managed by the FlightIntegrator should be and would make their orbits deviate from what would be expected, so predicting their positions using any method other than what KSP is using would quickly become inaccurate.
This is also apparent with the example of a craft under constant acceleration (plus gravity), which seems to be inconsistent with leap-frog but reasonably consistent with semi-implicit Euler, up to whatever correction the Orbital Drift Compensation is applying, which seems to be small enough in most cases as to be within the cone of fire from a gun with non-zero deviation, so it may not be worth worrying about.
However, since the bullets are not managed by the FlightIntegrator, but rather by our leap-frog implementation, this doesn't apply to them.

In some testing last night, I also noticed that `AIUtils.PredictPosition` gives quite bad values for predicting where a craft should be when high speeds are involved over longer times (e.g., it looked fine for 2s, but was off by several 10s of metres for 20s), which is not too surprising since under orbital conditions the acceleration is not constant over that period. The error from this should grow quadratically over time (I haven't had time to actually check this, but it's what's expected from the difference between an orbit and the quadratic trajectory kinematics predicts for constant acceleration).

Combining these observations, I think the best way forward with this is to replace the predicted position of the target in `BallisticTrajectoryClosestApproachSimulation` with the simulated position using semi-implicit Euler, keep the simulation of the position of the bullets using leap-frog and use these two positions to make the correction to `finalTarget`. This would be something like
```csharp
do
{
  ...
  timeToCPA = AIUtils.TimeToCPA(bullet relative);
  targetPredictedPosition = AIUtils.PredictPosition(target);
  finalTarget = targetPredictedPosition + bulletDropOffset - part.rb.velocity * timeToCPA;
  ...
} while(...);
if (...) // Expand this condition to include appropriate check for orbital conditions
{
  bulletCPA, targetCPA = BallisticTrajectoryClosestApproachSimulation(...);
  correction = bulletCPA - AIUtils.PredictPosition(bullet) + targetCPA - AIUtils.PredictPosition(target); // These are corrections to the positions in the do..while loop.
  finalTarget -= correction;
}
```

JacekCzupyt commented:
I've tried to research how exactly OBC (Orbital drift compensation) works, without much luck. Found some values related to it in PhysicsGlobals, but they don't seem very helpful:
```csharp
configNode.SetValue("orbitDriftFramesToWait", this.orbitDriftFramesToWait.ToString(), "Number of frames to wait once drift error threshold is met before drift compensation turns on", true);
configNode.SetValue("orbitDriftSqrThreshold", this.orbitDriftSqrThreshold.ToString(), "Square of the magnitude of the position error vector to use as threshold for drift compensation. If the error in position between the current position and the calculated rails position is < this, then drift compensation will be engaged.", true);
configNode.SetValue("orbitDriftAltThreshold", this.orbitDriftAltThreshold.ToString(), "Orbit radius threshold for drift compensation. If the orbital radius is less than this, then drift compensation will be engaged.", true);
```

I came up with this integrator which experimentally appears the well approximate the FlightIntegrator. It's essentially a combination of implicit-euler for thrust acceleration and leapfrog for gravitational acceleration.
```csharp
// Hybrid-integrator (no clue)
v3 += (vessel.acceleration_immediate - 0.5f * vessel.graviticAcceleration) * Time.fixedDeltaTime; // Update velocity by full step of manual acceleration and half step of gravitatinoal acceleration
x3 += v3 * Time.fixedDeltaTime; // Update the position
v3 += 0.5f * vessel.graviticAcceleration * Time.fixedDeltaTime; // Update velocity another half a step of gravitatinoal acceleration
```

Log results:
```
//not thrust 2s

DEBUG After 2s, Actual x = (-0.6, 3.2, -0.1), 
Forward Euler predicts x = (-0.6, 3.2, 0.0) (Δ = 0.1361977), 
Semi-implicit Euler predicts x = (-0.6, 3.2, -0.2) (Δ = 0.1361962), 
Leap-frog predicts x = (-0.6, 3.2, -0.1) (Δ = 0.0005098857), 
Hybrid predicts x = (-0.6, 3.2, -0.1) (Δ = 0.0005098857)

//5g 2s

DEBUG After 2s, Actual x = (2.0, -1.0, 2.3), 
Forward Euler predicts x = (2.7, -1.3, 3.3) (Δ = 1.278561), 
Semi-implicit Euler predicts x = (2.0, -1.0, 2.2) (Δ = 0.1365223), 
Leap-frog predicts x = (2.4, -1.1, 2.8) (Δ = 0.5891909), 
Hybrid predicts x = (2.0, -1.0, 2.3) (Δ = 0.005576788)

//15g 20s

DEBUG After 20s, Actual x = (0.0, 0.0, 0.0), 
Forward Euler predicts x = (55.9, -1.3, 10.8) (Δ = 56.96691), 
Semi-implicit Euler predicts x = (0.1, 0.0, -1.4) (Δ = 1.38957), 
Leap-frog predicts x = (28.0, -0.6, 4.7) (Δ = 28.41378), 
Hybrid predicts x = (-0.2, 0.0, 0.0) (Δ = 0.1618719)

//no thrust 20s

DEBUG After 20s, Actual x = (1.9, -1.1, 2.4), 
Forward Euler predicts x = (1.4, -1.0, 3.7) (Δ = 1.391746), 
Semi-implicit Euler predicts x = (2.2, -1.1, 1.0) (Δ = 1.395772), 
Leap-frog predicts x = (1.8, -1.1, 2.3) (Δ = 0.1130209), 
Hybrid predicts x = (1.8, -1.1, 2.3) (Δ = 0.1130209)
```

---

I've made a separate debug test for testing position predictors, and implemented a simulation based predictor to try to apprximate the FlightIntegrator:
```csharp
/// <summary>
/// Simulate a future position of an object given its current position, velocity and acceleration,
/// accounting for change in gravitational force and FlightIntegrator errors.
/// Assume provided acceleration is a sum of gravitational and some constant acceleration
/// </summary>
/// <param name="position">Initial position</param>
/// <param name="velocity">Initial velocity</param>
/// <param name="acceleration">Initial acceleration</param>
/// <param name="time">Predict state at time</param>
/// <param name="timeStep">Simulation time step</param>
/// <param name="isRb">Object is rigid body</param>
/// <returns>(position, velocity, acceleration) at requested time</returns>
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static (Vector3, Vector3, Vector3) PredictStateBallistic(Vector3 position, Vector3 velocity, Vector3 acceleration, float time, float timeStep, bool isRb = true)
{
    float timeElapsed = 0;
    // Get constant acceleratoin independent of gravity
    Vector3 a0 = acceleration - FlightGlobals.getGeeForceAtPosition(position);
    
    // Correct integration error
    if (isRb) 
    {
        velocity += a0 * Time.fixedDeltaTime / 2;
        
        // TODO: where to find ODC setting state?
        // if(ORBITAL_DRIFT_COMPENSATION == false)
        //     velocity += FlightGlobals.getGeeForceAtPosition(position) * Time.fixedDeltaTime / 2;
    }

    // First leapfrog initial step
    velocity += acceleration * timeStep / 2;
    
    // Leapfrog
    while (timeElapsed < time && timeStep > Time.fixedDeltaTime)
    {
        while (timeElapsed + timeStep > time && timeStep > Time.fixedDeltaTime)
        {
            timeStep /= 4;
            velocity -= (a0 + FlightGlobals.getGeeForceAtPosition(position)) * timeStep * 3 / 2;
        }

        if (timeStep <= Time.fixedDeltaTime)
        {
            velocity -= (a0 + FlightGlobals.getGeeForceAtPosition(position)) * (timeStep - time + timeElapsed) / 2;
            timeStep = time - timeElapsed;
        }

        position += velocity * timeStep;
        velocity += ((a0) + FlightGlobals.getGeeForceAtPosition(position)) * timeStep;
        timeElapsed += timeStep;
    }
    velocity -= (a0 + FlightGlobals.getGeeForceAtPosition(position)) * timeStep / 2;
    
    // Correct integration error
    if (isRb) 
    {
        velocity -= a0 * Time.fixedDeltaTime / 2;
        
        // TODO: where to find ODC setting?
        // if(ORBITAL_DRIFT_COMPENSATION == false)
        //     velocity -= FlightGlobals.getGeeForceAtPosition(position) * Time.fixedDeltaTime / 2;
    }
    
    return (position, velocity, a0 + FlightGlobals.getGeeForceAtPosition(position));
}
```
It is essencially just the same leapfrog used in ClosestApproachSimulation, that offsets the initial velocity to try to correct for the FlightIntegrator errors.
Log results:
```
// no thrust, 900k semi-major axis, 0.2 inclination

DEBUG After 20s, with a 5.12s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-2.1, 3.2, -0.9) (Δ = 1.697228)
DEBUG After 20s, with a 5.12s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-2.1, 3.2, -0.8) (Δ = 1.704123)

DEBUG After 20s, with a 2.56s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-0.9, 3.2, -0.5) (Δ = 0.4735709)
DEBUG After 20s, with a 2.56s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-0.9, 3.2, -0.5) (Δ = 0.469873)

DEBUG After 20s, with a 1.28s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-0.8, 3.2, -0.4) (Δ = 0.2719083)
DEBUG After 20s, with a 1.28s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-0.7, 3.2, -0.4) (Δ = 0.1645454)

DEBUG After 20s, with a 0.64s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-0.5, 3.2, -0.4) (Δ = 0.1263098)
DEBUG After 20s, with a 0.64s timestep, Actual x = (-0.5, 3.2, -0.3), 
Hybrid predictor predicts x = (-0.6, 3.2, -0.3) (Δ = 0.06955412)

// missile under 15g (same starting state)
DEBUG After 20s, with a 5.12s timestepn, Actual x = (0.0, 0.0, 0.0), //prograde 
Hybrid predictor predicts x = (-1.9, -0.1, -0.8) (Δ = 2.116998)
DEBUG After 20s, with a 5.12s timestepn, Actual x = (0.0, 0.0, 0.0), //retrograde
Hybrid predictor predicts x = (-1.0, 0.0, -0.3) (Δ = 0.9985156)

DEBUG After 20s, with a 2.56s timestepn, Actual x = (0.0, 0.0, 0.0), 
Hybrid predictor predicts x = (-0.7, 0.0, -0.5) (Δ = 0.8605039)
DEBUG After 20s, with a 2.56s timestepn, Actual x = (0.0, 0.0, 0.0), 
Hybrid predictor predicts x = (-0.3, 0.0, -0.1) (Δ = 0.2688451)

DEBUG After 20s, with a 1.28s timestepn, Actual x = (0.0, 0.0, 0.0), 
Hybrid predictor predicts x = (-0.3, 0.0, -0.2) (Δ = 0.3505023)
DEBUG After 20s, with a 1.28s timestepn, Actual x = (0.0, 0.0, 0.0), 
Hybrid predictor predicts x = (0.0, 0.0, 0.0) (Δ = 0.01863581)

DEBUG After 20s, with a 0.64s timestepn, Actual x = (0.0, 0.0, 0.0), 
Hybrid predictor predicts x = (-0.3, 0.0, -0.2) (Δ = 0.3136431)
DEBUG After 20s, with a 0.64s timestepn, Actual x = (0.0, 0.0, 0.0), 
Hybrid predictor predicts x = (-0.1, 0.0, 0.0) (Δ = 0.0563563)
```

At a timestep of 2.56s or 1.28s this seems good enough for any practical purpuses. I'll try to replicate this algorithm in the simulation based correction part of the aiming script.

Here is a patch with the tests and algorithm: 
[Ballistic_target_prediction.patch](https://github.com/BrettRyland/BDArmory/files/12383245/Ballistic_target_prediction.patch)

PS: I haven't been able to find a way to check if ODC is currently active, the PhysicsGlobals values i mentioned don't change. Also they suggest they ODC doesn't work at some altitues or some conditions, because apparently the KrakensBane mess isn't bad enough. At least the effect of this is very small.

BrettRyland commented:
I haven't had much time to look at this the last few days, however I have been experimenting a little bit and have some minor corrections and adjustments to the kinematic smoothing of the target's velocity and acceleration in `SmoothTargetKinematics` (this is particularly needed if the target is landed as the acceleration jumps around a lot due to the corrections for being on the surface).
There also seems to be some minor offsets due to the relative altitude between the firing craft and the target even when perfectly stationary, which I believe is due to how the final target is adjusted for the "bullet drop". This will likely be aggravated when in orbit due to the likelyhood of firing between significantly different altitudes.
I'll push these changes to a separate branch in a bit so as not to affect dev until this is working properly.

For the orbital drift compensation, based on the descriptions here
```
configNode.SetValue("orbitDriftFramesToWait", this.orbitDriftFramesToWait.ToString(), "Number of frames to wait once drift error threshold is met before drift compensation turns on", true);
configNode.SetValue("orbitDriftSqrThreshold", this.orbitDriftSqrThreshold.ToString(), "Square of the magnitude of the position error vector to use as threshold for drift compensation. If the error in position between the current position and the calculated rails position is < this, then drift compensation will be engaged.", true);
configNode.SetValue("orbitDriftAltThreshold", this.orbitDriftAltThreshold.ToString(), "Orbit radius threshold for drift compensation. If the orbital radius is less than this, then drift compensation will be engaged.", true);
```
and the values in GameData/ModuleManager.Physics:
```
orbitDriftFramesToWait = 5
orbitDriftSqrThreshold = 1E-10
orbitDriftAltThreshold = 400000000
```
I think that orbital drift compensation should only be being applied after 5 frames when the difference between the craft's position and the on rails value of its orbit drifts by more than 1e-5 when it's under 400Mm. This would imply that it's only applied while the craft isn't thrusting (but no guarantees since this is KSP spaghetti code).

JacekCzupyt commented:
Update: I've modified my code with orbital drift compensation in mind, and I'm getting some weird results. On one hand sometimes it works great, managing to hit a consistent stream of bullets on an 80km (impact time) target, 80s bullet flight time, at different orbital elevations, and 230m/s relative velocity. 

However, sometimes something in the code or ksp switches, and the guns suddenly start to very consistently miss (both long range oribtal and close range high acceleration targets) for up to minutes at a time. This seems to persist even through save loads, before it just as randomly becomes perfectly accurate again.

I'm not sure if this is new behaviour, or I just didn't notice it before. My best guesses are some error in iTime calculations, or maybe some race condition between two fixed update scripts. Investigating it now.

On a separate note, a lot of things still don't work well in orbit under 100km because of KrakensBane, not just aiming but also tracers and relative bullet checks.

BrettRyland commented:
I'm not surprised. I've found a few bugs that would significantly affect the aiming calculations. In particular,
- Smoothing of the target velocity and acceleration wasn't compensating for Krakensbane shifts. This smoothing is required due to spurious acceleration values of landed targets for both acceleration and acceleration_immediate (similar to the orbital drift compensation, but more dramatic since the vessel has to stay in contact with the terrain).
- The cubic solver was giving incorrect solutions for near-degenerate cubics due to tolerance checks in some situations. This was severely affecting near-zero relative acceleration situations.

I've fixed these and added some other minor corrections and pushed it to a new branch "aiming-again". Note that it's still a work in progress though.

I have a simple craft for testing accuracy [Aim tester.craft](https://github.com/BrettRyland/BDArmory/files/12437604/Aim.tester.craft.txt) (it needs infinite ammo/fuel/EC enabled) that can pull up to 10G and a custom turret with no bullet deviation, no recoil, 200km range and 1km/s bullets [50calsniper.cfg](https://github.com/BrettRyland/BDArmory/files/12437603/50calsniper.cfg.txt) (add this to the 50CalTurret folder).
With a pair of these at the limit of Kerbin's SoI (~82Mm), the gravitational acceleration is negligible, which allows testing the analytic solution.
However, I'm still getting some unexplained behaviour, such as requiring an extra shift of around `part.rb.velocity * Time.fixedDeltaTime` to the `finalTarget` when viewed from the target such that `part.rb.velocity` is non-zero in the Krakensbane velocity frame and it looks like the acceleration isn't quite being compensated for either, e.g., with a separation of around 20km and the target craft accelerating back and forwards at >1G, the shots tend to lag behind the target.
Since the numerical simulation is meant to correct for the deviation from the analytic solution, if the amount being added to `finalTarget` contains inaccuracies not related to the difference between the analytic and simulated trajectories, then it will also be off by a similar margin with the correction from the simulation added in.

JacekCzupyt commented:
College stuff caught up to me, so I didn't have much time in the last few weeks, sorry.

I've merged the changes from aiming-again with my own. I've been testing on edge of the Kerbin system, so I haven't tested the performence in a changing gravitatinoal field yet. I've gotten the medium range and high acceleration situations to work though. At long ranges I'm getting seemingly reandom innacuries of around 1-2m per 50km (they don't appear to be affected by relative velocity or acceleration), which I'm still investigating, I don't think floating point errors should be that big there.
Some clarifications:
1. The extra shift of around `part.rb.velocity * Time.fixedDeltaTime` to the `finalTarget` is due to the initial frame we skip, the one the bullet is spawned on. This is already compensated for in the `targetPredictedPosition`, and we don't need it for the `bulletDropOffset`, since the bullet doesn't experience gravity during the frame it is spawned on.
2. The shot was lagging behind the target because the analytic solution on the `aiming-again` branch was not accounting for the error caused by the semi-implicit euler used by unity.
3. The `BallisticTrajectoryClosestApproachSimulation` on the `aiming-again` branch appears to be using semi-implicit euler for sumilating the target. We can't do that, because we are using a much bigger timestep then `Time.FixedDeltaTime` for that simulation, so we will get completly different (much bigger) errors then the unity integrator. 
Instead we can leverage the fact that for a constant acceleration, these errors are consistent and predictable. During each fixed frame, the error will be `constTargetAcceleration * Time.FixedDeltaTime * Time.FixedDeltaTime / 2`, which means `constTargetAcceleration * Time.FixedDeltaTime / 2` per second. If we split the initial targetAcceleration into gravitational and constant, we can simply add this error after each timestep.
Instead of doing that manually, I'm adding an initial offset to the targets velocity in line 3534:
https://github.com/JacekCzupyt/BDArmory-Orbital/blob/bae32cab1dc855ef054ade44940b18af59318df8/BDArmory/Weapons/ModuleWeapon.cs#L3534
This also corrects for the problem in the analytic solution.

PS. Should I change the pull request target branch to `aiming-again`?

BrettRyland commented:
No worries. I've been quite busy with other stuff too lately and haven't had much time to work on this either. Hopefully, I'll have a bit more time to look at it this week.

In order to get the aiming code working correctly, there are essentially three situations that the aiming code has to handle:
- Static: where the two vessels are supported on or near the surface of Kerbin.
    - In this situation, there is no effect from velocity, acceleration or Krakensbane, just the variation in gravity over the path of the bullet.
    - BDA's Vessel Mover gives perfectly stationary vessels, and the kinematic smoothing should give sufficiently static landed/splashed vessels for testing against.
    - The numerical integrator for the target is irrelevant here due to the target being static.
    - This situation is useful for getting the initial setup of the bullets and their trajectories (e.g., bullet drop, iTime) correct. Bullets should closely follow the debug lines.
- Translational: where variation in gravity is negligible, e.g., at the limit of Kerbin's SoI.
    - In this situation, the solution should be analytically solvable for constantly accelerating vessels once changes in the Krakensbane velocity offloading are accounted for.
    - This situation is useful for getting the velocity corrections to finalTarget correct (e.g., part.rb.velocity) and any other peculiarities of the analytic solution.
    - The solution from numerical simulation should agree closely with the analytic solution here — can be used to partially validate the numerical accuracy of KSP/Unity (only partially since the simplicity of the Hamiltonian may mean that the integrators appear more accurate here than they would normally be).
    - This situation also covers most short-range conditions during in atmosphere dogfights, which generally are sufficiently covered by the analytic solution.
- Orbital (>100km): where varying gravity, vessel acceleration and Krakensbane must all be accounted for.
    - In this situation, the numerical integrators for bullets and the target need to closely approximate their actual trajectories.
    - This situation is useful for making sure that the way finalTarget is calculated works in this geometry (i.e., is combining bulletDrop and part.rb.velocity to get the firing direction sufficient or are they just the lowest order terms when in orbit?) and dealing with KSP's orbital drift compensation (separate gravitational vs local acceleration).
    - For <100km orbits, corrections due to Krakensbane may need adjusting.

There are still some bugs in the current implementation that I'm trying to sort out. Something changed between the version in dev and what I have locally in the aiming-again branch (which contains some un-committed stuff) that causes a significant offset in the static situation, which I need to fix first. I.e., in the dev branch the "aim tester" craft can precisely hit another one at 100km (both suspended at 10km altitude above Kerbin), but in the aiming-again branch it's aiming high for some reason.

Before I noticed this issue with the static situation, I had something mostly working for the translational situation, but it required an unexplained correction to the analytic solution (within the do-while loop) of approximately
```
finalTarget += 0.5f * Time.fixedDeltaTime * timeToCPA * relativeAcceleration
```
which, together with the correction to the `targetVelocity` due to Unity's integrator that you're proposing, seems to be fairly accurate (e.g., hitting at 9G acceleration at 20-50km distances with 3-5km/s speed differences), but I don't like having unexplained corrections.

I'm also seeing the semi-random inaccuracies at very long ranges too. At long distances, when the target vessel has the focus, the velocity of the turret part is oscillating a small amount (likely due to Krakensbane), which might be causing this. I'll probably need to add smoothing to that part's velocity too.

It's probably a good idea to change the PR to target the aiming-again branch, yeah.

JacekCzupyt commented:
In a translational situation, your correction
`finalTarget += 0.5f * Time.fixedDeltaTime * timeToCPA * relativeAcceleration`
is basically completly equivalent to adding the line
`targetVelocity += 0.5f * Time.fixedDeltaTime * relativeAcceleration` 
before the analytic solution, which is what I'm using to correct for the unity integrator, so that's almost certainly what it was correcting for.

BrettRyland commented:
> In a translational situation, your correction `finalTarget += 0.5f * Time.fixedDeltaTime * timeToCPA * relativeAcceleration` is almost completly equivalent to adding the line `targetVelocity += 0.5f * Time.fixedDeltaTime * targetAcceleration` before the analytic solution, which is what I'm using to correct for the unity integrator, so that's almost certainly what it was correcting for.

That's what I initially thought too, but experimentally, the `targetVelocity += ...` line makes a much smaller difference than the `finalTarget += ...` line and the combination of the two seems to be more accurate. So I suspect there's still an off-by-one-frame mistake somewhere in the calculations.

BrettRyland commented:
OK, I've found out what broke with the static situation and it's going to require a small rework of how aiming is done from frame-to-frame since it will also affect the orbital situation.

The `do...while` loop converges to the solution by roughly an order of magnitude per loop, which is fine if the non-analytic correction isn't needed. In particular, `bulletEffectiveVelocity` is only changed to use a `firingDirection` other than `fireTransforms[0].forward` if more than 1 iteration is performed.
However, if the non-analytic correction is needed, then it's only performing a single iteration (for performance reasons) of this convergence with the `bulletEffectiveVelocity` from the analytic solution, which can be several hundred metres off-target at long ranges on Kerbin, so the single iteration doesn't bring it particularly close to the correct solution unless `bulletEffectiveVelocity` is already close to what it's supposed to be. For the static situation, `bulletEffectiveVelocity` converges over several frames if only a single iteration occurs in the `do..while` loop, but for the orbital situation it might not fully converge due to the correct solution changing over time.

The way I intend to fix this (once I have time to do so) is to track the firing direction (and anything else that's needed) from frame to frame so that the aiming should only need to perform a single iteration based on the previous `firingDirection` unless the weapon switches targets.

JacekCzupyt commented:
Yup, I've already got that covered in my PR
https://github.com/JacekCzupyt/BDArmory-Orbital/blob/bae32cab1dc855ef054ade44940b18af59318df8/BDArmory/Weapons/ModuleWeapon.cs#L3600-L3603

I would not be surprised if there were still some off-by-one frame errors of varrying severity in the calculations, but the fact that those two lines of code yield noticably different results is confusing, not sure what would cause that. Maybe something to do with KrakensBane? I'll investigate that.

I've started testing the orbital correction in LKO (~200 km) again, there's an offset of around 1m per 30km, so I probably messed something up. Also for some reason when in LKO, in high acceleration tests the bullets are lagging behind the target at close ranges (3-5km), even when simulation correctoins are turned off. This disapears when i switch to the fireing vessel, so it's almost certainly KrakensBane related, but I don't know why I can't replicate this in the translational tests at the edge of the Kerbin system.

BrettRyland commented:
> Yup, I've already got that covered in my PR https://github.com/JacekCzupyt/BDArmory-Orbital/blob/bae32cab1dc855ef054ade44940b18af59318df8/BDArmory/Weapons/ModuleWeapon.cs#L3600-L3603

No, that doesn't fix what I was talking about. `firingDirection` _should be_ `fireTransforms[0].forward` so that the time to the closest point of approach is based on where the bullet is actually going when the weapon is roughly pointing in the right direction, otherwise the aiming direction converges (over time) to a direction as if `fireTransforms[0].forward` were pointing in the modified `firingDirection` even though it isn't. The reason for the `do...while` loop with the modified `firingDirection` is that when the weapon is _not_ pointing in roughly the right direction, then the `TimeToCPA` calculation gives poor (or simply wrong if pointing away from the target) results and iterating the loop allows the calculation to converge to the right _analytic_ value _as if the weapon were pointing in the right direction_. This is also why the angle check for the `do...while` loop `shouldn't` be too tight. It should only get `fireDirection` pointing _roughly_ in the right direction so that the turret doesn't fire prematurely due to thinking it's on target while the aiming direction is still converging (which was something that happened sometimes).
Resetting `fireDirection` to `fireTransforms[0].forward` for the numeric solution helps somewhat (at least for the static situation), but leads to the correction being larger than it needs to be to compensate for the incorrect `fireDirection` from the extra loops in the analytic solution. This is going to be exacerbated in the orbital situation.

> Edit: After rereading, im not sure I understand the fix you proposed, are you saying you indend to remove the `do...while` loop and only execute the analytical aiming iteration once per frame? If so, I'm not sure that it would work well fast moving targets when the angle to target shifts significantly frame by frame. Also if the weapon was already close to pointing at the target (i.e. when precision starts to matter) then `fireTransforms[0].forward` should be basically equal to `firingDirection` from the previous frame, right? Maybe the problem comes from the "Rough initial estimate" of the `timeToCPA`?

Yes, in what I'm proposing to do, it would run either the analytic solution or the numeric CPA only once per frame as long as we have a reasonable estimate of `finalTarget`, which double exponential smoothing should give. When changing targets or if the weapon is significantly off-target, then it'll probably still need to run the `do...while` loop for a few frames until the `finalTarget` estimate stabilises.

BrettRyland commented:
Hey, I've been working on this a bit lately. Progress has been somewhat slow and tedious, but I believe the code is working fairly well for the first two situations now.
For the "orbital" situation, however, the aim is still a bit off when the craft are separated by a large distance in low Kerbin orbit (100-200km), though it usually hits when the craft are within 5-10km.

I suspect that the reason leap-frog is working so much better than semi-implicit Euler in the "translational" situation is due to the acceleration of the target vessel being constant (meaning the potential in the Hamiltonian is quadratic) and the leap-frog method being second order, so that it's effectively solving it exactly.
In the "orbital" situation, the acceleration isn't constant (it'd be a perturbation of the quadratic potential with probably cubic terms), so the solution from leap-frog is a little off (either that or there's a bug in the `targetGravity` parts of the simulation). The orbital drift compensation wouldn't be helping either, but I would think that that should only be being applied when a craft isn't thrusting (it wouldn't make sense to apply while the craft is thrusting since the orbit is being changed every frame, but this is KSP...).
It's possible an adjustment to the `targetGravity` part of the simulation would be enough to compensate for the aiming error here, but it will need some more experimentation.

JacekCzupyt commented:
That mostly makes sense, when experimenting with the integrators in LKO, the best ones i testsed generally got an error of around 1m in 60s tests with a 128*FixedDeltaTime timestep, and increased more or less quadratically. Even if everything else is perfect, this might put a limit on distance (or time to target) in these scenarios. I'm happy that close range situations have improved, as that was the reason I originally started working on this.

Unfurtunetly I don't currently have any time to work on this due to work and collage, I hope to return to it once I graduate at the end of the year.

BrettRyland commented:
With the latest change in dev from commit b742bd12a, the orbital situation is even closer to being correct now. It's still off slightly at longer ranges (20-30km for 1km/s bullets), but ought to be within the max deviation of most guns.
